core/arch/arm/tee/entry_std.c

```diff
@@ -350,21 +514,31 @@ void tee_entry_std(struct thread_smc_args *smc_args)
 		return;
 	}
 	parg = (uint64_t)smc_args->a1 << 32 | smc_args->a2;
-	if (!tee_pbuf_is_non_sec(parg, sizeof(struct optee_msg_arg)) ||
-	    !ALIGNMENT_IS_OK(parg, struct optee_msg_arg) ||
-	    !(arg = phys_to_virt(parg, MEM_AREA_NSEC_SHM))) {
-		EMSG("Bad arg address 0x%" PRIxPA, parg);
-		smc_args->a0 = OPTEE_SMC_RETURN_EBADADDR;
-		return;
+
+	/* Check if this region is in static shared space */
+	if (core_pbuf_is(CORE_MEM_NSEC_SHM, parg,
+			  sizeof(struct optee_msg_arg))) {
+		mobj = get_cmd_buffer(parg);
+	} else {
+		if (parg & SMALL_PAGE_MASK) {
+			smc_args->a0 = OPTEE_SMC_RETURN_EBADADDR;
+			return;
+		}
+		mobj = map_cmd_buffer(parg);
 	}
 
-	num_params = arg->num_params;
-	if (!tee_pbuf_is_non_sec(parg, OPTEE_MSG_GET_ARG_SIZE(num_params))) {
+	if (!mobj || !ALIGNMENT_IS_OK(parg, struct optee_msg_arg)) {
 		EMSG("Bad arg address 0x%" PRIxPA, parg);
 		smc_args->a0 = OPTEE_SMC_RETURN_EBADADDR;
+		mobj_free(mobj);
 		return;
 	}
 
+	arg = mobj_get_va(mobj, 0);
+	assert(arg && mobj_is_nonsec(mobj));
+
+	num_params = arg->num_params;
```

Comment:
We should either use the value previously read by `get_cmd_buffer()` or `map_cmd_buffer()`, or we'll need to check that the value we read now is small enough for the `mobj`. I'd prefer to only read once to keep it simple.